share/picongpu/tests/CurrentDeposition/lib/python/test/CurrentDeposition/grid_class.py

```diff
@@ -0,0 +1,207 @@
+
+        start_coord = np.array([start_x, start_y, start_z])
+
+        # Calculate end coordinates - starting from starting cell, add position
+        # and any cell crossings pos_off1 - pos_off2
+        end_x = start_x_cell + pos2[0] - pos_off1[0] + pos_off2[0]
+        end_y = start_y_cell + pos2[1] - pos_off1[1] + pos_off2[1]
+        end_z = start_z_cell + pos2[2] - pos_off1[2] + pos_off2[2]
+
+        end_coord = np.array([end_x, end_y, end_z])
+
+        return start_coord, end_coord
+
+    def current_deposition_field(self, start_coord, end_coord, grid_x,
+                                 grid_y, grid_z):
+        """Calculates the current deposition vector component-wise for z, y, x.
+        Returns the components of the vector in the x, y, z direction on the
+        grid. Takes the tuple of start and end coordinates (x, y, z) as well
+        as the zero grids for each coordinate.
+        """
+        order = self.order
+        func = self.func
+        num_cells = order + 3 + 1
+
+        for z in range(num_cells):
+            for y in range(num_cells):
+                for x in range(num_cells):
+                    # Calculate the si components for the W calculation
+
+                    # Old
+                    s1 = func(start_coord[0] - x)
+                    s2 = func(start_coord[1] - y)
+                    s3 = func(start_coord[2] - z)
+                    # new
+                    s4 = func(end_coord[0] - x)
+                    s5 = func(end_coord[1] - y)
+                    s6 = func(end_coord[2] - z)
+
+                    # calculate W and assign it to the grid node
+                    grid_x[z][y][x] = W(s1, s2, s3, s4, s5, s6)
+                    grid_y[z][y][x] = W(s2, s1, s3, s5, s4, s6)
+                    grid_z[z][y][x] = W(s3, s2, s1, s6, s5, s4)
+
+        return grid_x, grid_y, grid_z
+
+    def current_density_field(self, W_grid_x, W_grid_y, W_grid_z, start_coord,
+                              end_coord, Q, cell_depth, cell_width,
+                              cell_height, dt, unit_time, unit_charge,
+                              unit_length):
+        """Calculates the current density in the x, y, and z directions on the
+        grid using the current deposition vector(field) W_grid_x,y,z. Charge Q
+        is needed as input. dx, dy, dz, and dt are set to 1 by default.
+        """
+
+        order = self.order
+        num_cells = order + 3 + 1
+        c = 299792458.0
+
+        # create null grid to write current density to
+        # due to periodicity, the grid has the same shape -> i = 1/2 is grid[0]
+        null_axis = np.zeros(num_cells, dtype=float)
+        j_grid_x, j_grid_y, j_grid_z = np.meshgrid(null_axis, null_axis,
+                                                   null_axis)
+
+        # determine how many grid points need to be included in calculation
+        # aka half-width of assignment function
+        # coordinate of coordinate to be calculated

```

Comment:
'coordinate of coordinate to be calculated'?

Reply:
removed